### ISSUE TITLE
mtTkinter prevents sys.exit() from executing, leaving processes running (Because some threads remain open?)

### ISSUE THREAD

**Issue by casparwylie** [RedFantom/mtTkinter#1]

**Comment by RedFantom** [comment]
Before ending a Tkinter application, the `destroy()` function should be called first in order to end the `mainloop`. Also, mtTkinter does not actually create any threads to run in, it just uses `Queue` objects to handle any event calls from other threads.

Of course, there  may still be a bug. Can you provide a verifiable example (preferably as small as possible) of a program in which you see this behaviour? You could also run `print(threading.enumerate())` to check what threads are still open at the end of your program.

**Comment by casparwylie** [comment]
I do destroy the mainloop before hand.  I had a look at the threads running just before attempt to exit, and it only says the mainThread is, which is weird. When I use normal Tkinter, with no other changes, it works fine.

**Comment by RedFantom** [comment]
If you look at the `_CheckEvents(tk)` function in `mtTkinter.py`, you can see that the function will quit its inifinite loop on one of two conditions:

- A `SystemExit` is raised
- There are no more events in the `_eventQueue`

While I do not know the internals of the `sys.exit` function, I do know that if you call the `destroy` function, the `mainloop` ends and there *should* not be any events left in the queue. Again, could you provide an example of the issue you're experiencing?

**Comment by casparwylie** [comment]
You're right. I am struggling to reproduce a simplified version of issue, which suggests something is possibly not right on my end.  Will get back to you in a few hours.

**Comment by baldk** [comment]
I have been struggling with this issue as well and it took me quite a while to figure out that it was the wrapper, i.e. mtTkinter that was the culprit and not Tkinter. Nevetheless I have put some code together that will reproduce the issue.

```python
from mttkinter import mtTkinter as tk

def window_delete_handler():
	root.destroy()
	
root = tk.Tk()
root.protocol("WM_DELETE_WINDOW", window_delete_handler)
labelVar = tk.IntVar(root, value=0)

root.mainloop()
```

If the Variable is removed from the code it does not hang.

```python
from mttkinter import mtTkinter as tk

def window_delete_handler():
	root.destroy()
	
root = tk.Tk()
root.protocol("WM_DELETE_WINDOW", window_delete_handler)

root.mainloop()
```

**Comment by casparwylie** [comment]
Yes - that's it. I was not able to reproduce because I was over simplifying (I did not add an` IntVal `or any other middle way Tkinter objects). So, in` window_delete_handler`,` sys.exit()` does not work correctly, however it does get called and raise `SystemExit`.

**Comment by baldk** [comment]
Getting closer...

Setting mtDebug=9 will show the last two lines before hang as

```python
Calling event directly: deletecommand ('42274392L_CheckEvents',) {}
Marshalling event: call ('info', 'exists', 'PY_VAR0') {}
```

So deleting the Variable after mainloop is done will make it exit.

```python
from mttkinter import mtTkinter as tk

def window_delete_handler():
	root.destroy()
	
root = tk.Tk(mtDebug=9)
root.protocol("WM_DELETE_WINDOW", window_delete_handler)
labelVar = tk.IntVar(root, value=0)

root.mainloop()
del labelVar
```

While this works it should be fixed in a better way.

Looking in Tkinter.py; the class Variable has the following member

```python
def __del__(self):
        """Unset the variable in Tcl."""
        if self._tk is None:
            return
        if self._tk.getboolean(self._tk.call("info", "exists", self._name)):
            self._tk.globalunsetvar(self._name)
        if self._tclCommands is not None:
            for name in self._tclCommands:
                #print '- Tkinter: deleted command', name
                self._tk.deletecommand(name)
            self._tclCommands = None
```

Maybe this needs to be improved in some way?

Update:

With the delete at the bottom mtDebug yields

```python
Calling event directly: globalunsetvar ('PY_VAR0',) {}
```

**Comment by casparwylie** [comment]
Interestingly deleting `IntVar` after mainloop does not work for me, either way, it wouldn't scale as a solution. It is worth remembering that if we replace mtTkinter with Tkinter, it all works fine. In mtTkinter, I don't think SystemExit ever gets raised.

**Comment by baldk** [comment]
Okay... getting even closer. We're not at root cause yet, but I do have a fix.

The offender is a final call coming from a non-creator thread (presumably the Queue as it is a DummyThread). A call from a non-creator thread is put into the eventQueue and a new responseQueue is created. The script will now sit there and wait for a response until one arrives which leads to system hang. Putting a get(False) i.e. nowait instead yields the response below, which indicates that the offender is indeed the Variable we created.

```python
Marshalling event: call ('info', 'exists', 'PY_VAR0') {}
Exception Queue.Empty: Empty() in <bound method IntVar.__del__ of <Tkinter.IntVar instance at 0x000000000277AC88>> ignored
```

So by ignoring calls from non-creator threads we can circumvent the problem and use mtTkinter as we have always used Tkinter. I have therefore made a fork of the code and issued a pull request. So if you can't wait you can go to the link below and download the fix.

[https://github.com/baldk/mtTkinter/tree/develop](url)

UPDATE:
This does not work. The non-creator threads are of course used when getting and setting the Variable. I will put in a timeout instead of complete ignore.

**Comment by RedFantom** [comment]
This is a much larger issue than I thought it was, and clearly this is not the desired way for the code to behave. Thank you both for looking into this! If either of you still feels up for it, I invite you to continue investigating the issue. I do not currently have time to do so, unfortunately, but if you are unable to too then I will try to find some time soon.

**Comment by baldk** [comment]
Added a new temporary solution for issue #1.

Not sure if it works all the time, but it does not require any flags during instantiation. Unfortunately I don't think I have more time to spare on this right now, so give it twirl and we'll see if it solves the issue.

**Comment by RedFantom** [comment]
This issue was closed automatically by GitHub once I referenced a fix to this issue from commit 5420252, but if either of you still wants to add something, feel free to!
